### PR TITLE
Allow for more than one .more message at a time

### LINKED ISSUE #232: `more` module: Allow users to request for a few messages at once
The `more` module as of now allows a user to request only one message at a time. Let the user be able to request multiple messages at once, by inputting a number.

### PR DESCRIPTION
**Pull request by Androbin**
Fixes #232 
@shardulc Friendly ping.

### CODE REVIEW

**Comment by shardulc** [comment]
Preferred behavior:
 - If `count` is provided and is greater than 1, the `X remaining messages` should be a separate message
 - Else, append `(X remaining)` to the single message that `.more` will output.

**Comment by shardulc** [comment]
@Androbin Also, since `.more` now has an optional argument, can you add a help message for the command? Something like [this](https://github.com/goavki/phenny/blob/master/modules/git.py#L333).

File: modules/more.py [hunk 1]
```diff
@@ -41,24 +41,46 @@ def add_messages(target, phenny, msg, break_up=break_up_fn):
         phenny.messages[caseless_nick] = msgs
 
 def more(phenny, input):
+    '''Print the (N) next messages. Possible commands are:
```
Comment (shardulc) [inline modules/more.py]: I suggest:
'.more N' prints the next N messages. If N is not specified, prints the next message.

File: modules/more.py [hunk 2]
```diff
@@ -41,24 +41,46 @@ def add_messages(target, phenny, msg, break_up=break_up_fn):
...
+            show_more(phenny, caseless_sender, count)
 
 more.name = 'more'
-more.rule = r'[.]more'
+more.rule = r'[.]more( ([1-9][0-9]*))?'
+
+def show_more(phenny, caseless_nick, count):
+    if count > 1:
+        for id in range(count):
+            msg = phenny.messages[caseless_nick][id]
```
Comment (shardulc) [inline modules/more.py]: Hmm... it prints the 0th message and then removes it, and then tries to print the 1st message when it should actually be printing the new 0th message, right? Also, what if `count` is greater than the number of messages?

Comment (wei2912) [reply modules/more.py]: The list should be treated as a queue, and entries popped off accordingly using the `.pop()` function for lists.

File: modules/more.py (hunk 1 above)
Comment (Androbin) [reply modules/more.py]: Okay

File: modules/more.py (hunk 2 above)
Comment (Androbin) [reply modules/more.py]: Sorry, messed that up.
Let me fix that.

File: modules/more.py [hunk 3]
```diff
@@ -41,24 +41,45 @@ def add_messages(target, phenny, msg, break_up=break_up_fn):
...
+
+    remaining -= count
+
+    if count > 1:
+        for _ in range(count):
+            phenny.reply(phenny.messages[caseless_nick].pop(0))
+
+        if remaining > 0:
+            phenny.reply(str(remaining) + " messages remaining")
+    else:
```
Comment (shardulc) [inline modules/more.py]: What if `remaining` is 0?

File: modules/more.py [hunk 4]
```diff
@@ -41,24 +41,45 @@ def add_messages(target, phenny, msg, break_up=break_up_fn):
...
+        msg = phenny.messages[caseless_nick].pop(0)
 
-def show_more(phenny, caseless_nick):
-    msg = phenny.messages[caseless_nick][0]
-    phenny.messages[caseless_nick].remove(phenny.messages[caseless_nick][0])
-    remaining = ' (' + str(len(phenny.messages[caseless_nick])) + ')' if phenny.messages[caseless_nick] else ''
-    phenny.reply(msg + remaining)
+        if remaining > 0:
+            phenny.reply(msg + " (" + remaining + " remaining)")
```
Comment (shardulc) [inline modules/more.py]: Should be `str(remaining)`

**Comment by shardulc** [comment]
@Androbin As you have done previously, can you write/update a test for this module? Be sure to cover all cases, otherwise something like my `str(remaining)` comment above might be missed.

File: modules/more.py (hunk 4 above)
Comment (Androbin) [reply modules/more.py]: Good catch.

**Comment by Androbin** [comment]
@shardulc I'll try my best to cover all cases.

**Comment by Androbin** [comment]
If there are no more messages to show, typing `.more` is a no-op.
It'd be nice to have it print something like "No stalled messages".

**Comment by shardulc** [comment]
"No more queued messages" might be better.

**Comment by Androbin** [comment]
Should this be printed by `more` or `show_more`?
I'm asking because of the `admin or owner` case.

**Comment by shardulc** [comment]
I think it should be printed in `show_more`. Instead of checking `if caseless_X in phenny.messages.keys():` in `more`, you could check it in `show_more` and reduce duplication.

**Comment by Androbin** [comment]
> Instead of checking `if caseless_X in phenny.messages.keys():` in `more`, you could check it in `show_more` and reduce duplication.

In the last PR I already had it there #irony. I'll put the return value back in.

**Comment by shardulc** [comment]
@Androbin No, the return value is a bad idea, because `show_more` then serves two confusing functions at once.

**Comment by Androbin** [comment]
But we still need to know that value in `more`.

**Comment by shardulc** [comment]
We don't. `more` can pass an argument to `show_more` like `admin_only=True` or something.

**Comment by Androbin** [comment]
But even admins and owners should get their personal messages first and moreover not intermixed with channel messages.

**Comment by shardulc** [comment]
I don't see where the confusion is. It is just a matter of moving the condition-checking into the `show_more` method, leaving the functionality the same.

**Comment by Androbin** [comment]
But it's an `if ... elif ...` which means the value is used here.

**Comment by Androbin** [comment]
Writing the tests is tricky due to various edge cases in `more` AND `add_messages`.

**Comment by Androbin** [comment]
Okay, I figure a way that should satisfy all sides:
* `more` doesn't handle `caseless`-ness itself either
* `has_more` encapsulates the logic
* `show_more` has no return value
* `has_more` and `show_more` handle `caseless`-ness

File: modules/more.py [hunk 5]
```diff
@@ -35,30 +37,49 @@ def add_messages(target, phenny, msg, break_up=break_up_fn):
...
-    elif input.admin or input.owner:
-        caseless_sender = input.sender.casefold()
+    count = 1 if input.group(2) is None else int(input.group(2))
 
-        if caseless_sender in phenny.messages.keys():
-            show_more(phenny, caseless_sender)
+    if has_more(input.nick):
+        show_more(phenny, input.nick, count)
+    elif (input.admin or input.owner) and has_more(input.sender):
+        show_more(phenny, input.sender, count)
```
Comment (shardulc) [inline modules/more.py]: Else show "no more messages", right?

File: modules/more.py [hunk 6]
```diff
@@ -35,30 +37,49 @@ def add_messages(target, phenny, msg, break_up=break_up_fn):
...
 
 more.name = 'more'
-more.rule = r'[.]more'
+more.rule = r'[.]more( ([1-9][0-9]*))?'
```
Comment (shardulc) [inline modules/more.py]: Use `r'[.]more(?:([1-9][0-9]*))'` and `input.group(1)`.

File: modules/test/test_more.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        self.input.admin = True
+        self.create_messages(self.input.nick, 5)
+        self.input.group = lambda x: [None, None, '3'][x]
+        more.more(self.phenny, self.input)
+
+        calls = [call(self.input.nick + ": " + message) for message in self.messages[1:4]]
+        calls.append(call("2 messages remaining"))
+        self.phenny.reply.assert_has_calls(calls)
+
+    def test_more_admin_only(self):
```
Comment (shardulc) [inline modules/test/test_more.py]: I suggest renaming these tests to `test_more_channel_admin_only` to make it clear that channel messages are being tested.

**Comment by shardulc** [comment]
@Androbin Nice work with the tests. This is almost ready to merge.

File: modules/more.py (hunk 5 above)
Comment (Androbin) [reply modules/more.py]: Totally forgot about that.

File: modules/more.py (hunk 6 above)
Comment (Androbin) [reply modules/more.py]: I was not sure if this syntax was supported.

File: modules/test/test_more.py (hunk 7 above)
Comment (Androbin) [reply modules/test/test_more.py]: Okay, gimme a moment.

**Comment by shardulc** [comment]
@Androbin Merged in b4db977. I realized you had added the `posted.db` file by accident, had to remove it.

**Comment by Androbin** [comment]
Just noticed this reverts 04b6f68d4e18dd641e276b9e13e91b0cb7b88c83.
